File: cpp-02/ex2/ft_custom_errors.cpp
#include <exception>
#include <iostream>
#include <string>
#include <vector>

class GardenError: public std::exception { };

class PlantError: public GardenError
{
    private:
        std::string message;
    public:
        PlantError(const std::string& msg): message(msg) { }
        const char* what() const noexcept
        {
            return message.c_str();
        }
};

class WaterError: public GardenError
{
    private:
        std::string message;
    public:
    WaterError(const std::string& msg): message(msg) { }
    const char* what() const noexcept
    {
        return message.c_str();
    }
};

void test_moisture(int moisture)
{
    if (moisture < 1)
        throw PlantError("The tomato plant is wilting!");
    else
        std::cout << "Plant OK!" << std::endl;
}

void test_water(int water)
{
    if (water < 1)
        throw WaterError("Not enough water in the tank!");
    else
        std::cout << "Water OK!" << std::endl;
}

void test_errors()
{
    try
    {
        std::cout << "Testing PlantError..." << std::endl;
        test_moisture(0);
    }
    catch (const PlantError& e)
    {
        std::cout << "Caught PlantError: '" << " " << e.what() << std::endl;
    }

    try
    {
        std::cout << "Testing WaterError..." << std::endl;
        test_water(0);
    }
    catch (const WaterError& e)
    {
        std::cout << "Caught WaterError: '" << " " << e.what() << std::endl;
    }

    std::vector<std::exception> exceptions;
    std::cout << "Testing all garden errors..." << std::endl;

    try
    {
        test_moisture(0);
    }
    catch (const PlantError& e)
    {
        exceptions.push_back(e);
    }

    try
    {
        test_water(0);
    }
    catch (const WaterError& e)
    {
        exceptions.push_back(e);
    }

    for(auto& i: exceptions)
        std::cout << "Caught a garden error: " << i.what() << std::endl;
    std::cout << "All custom error types work correctly!" << std::endl;
}

int main()
{
    test_errors();
    return 0;
}

File: python-02/ex2/ft_custom_errors.py
#!/usr/bin/env python3


class GardenError(Exception):
    pass


class PlantError(GardenError):
    def __str__(self):
        return "The tomato plant is wilting!"


class WaterError(GardenError):
    def __str__(self):
        return "Not enough water in the tank!"


def test_moisture(moisture):
    if moisture < 1:
        raise PlantError("The tomato plant is wilting!")
    else:
        print("Plant OK!")


def test_water(water):
    if water < 1:
        raise WaterError("Not enough water in the tank!")
    else:
        print("Water OK!")


def test_errors():
    try:
        print("Testing PlantError...")
        test_moisture(0)
    except PlantError as e:
        print(f"Caught PlantError: ' {e}")

    try:
        print("Testing WaterError...")
        test_water(0)
    except WaterError as e:
        print(f"Caught WaterError: ' {e}")

    exceptions = []
    print("Testing all garden errors...")

    try:
        test_moisture(0)
    except PlantError as e:
        exceptions.append(e)

    try:
        test_water(0)
    except WaterError as e:
        exceptions.append(e)

    for i in exceptions:
        print(f"Caught a garden error: {i}")
    print("All custom error types work correctly!")


if __name__ == "__main__":
    test_errors()
